### PR TITLE
Run replanning in its own thread (fixes #223)

### LINKED ISSUE #223: Canceling move_base action takes very long when doing continuous replanning
Looks like it takes approximately replanning period! All happens in this line:

```
  if(!action_client_get_path_.getState().isDone())
```
Most probably is caused by the calls to replanning_rate_.sleep(), so I suppose we must move the replanning to a separated thread.

### LINKED ISSUE #259: Actionlib bug when planner_frequency is anything other than 0
**ROS:** Noetic
**Package:** ros-noetic-move-base-flex 0.3.4-1focal.20210112.101859

**planners.yaml:**
```
planners:
  - name: 'planner'
    type: 'global_planner/GlobalPlanner'
```

I get the following console output when setting the global planner_frequency to anything other than 0hz:

```
BUG: Got a transition to CommState [ACTIVE] when in SimpleGoalState [DONE]
BUG: Got a second transition to DONE
```

The global planner only ends up publishing a total of 2 times /to `move_base_flex/planner/plan` and on the second time it outputs the above messages to the console.

This bug only happens only once per navigation goal. The robot does end up reaching its target however, and I get the `GOAL Reached!` message successfully.

### PR DESCRIPTION
**Pull request by corot**
As I mentioned in the issue, we need to run the replanning in its own thread

Interestingly, it also seems to fix #259 and #260 (both the same issue, I think)

@echo4echo, @r0gi, would be really great if you can give  a try to this branch and see if it fixes the issue you reported;  thanks in advance

### CODE REVIEW

File: mbf_abstract_nav/src/move_base_action.cpp [hunk 1]
```diff
@@ -45,22 +45,22 @@
...
+  , action_client_exe_path_(private_nh_, "exe_path")
+  , action_client_get_path_(private_nh_, "get_path")
+  , action_client_recovery_(private_nh_, "recovery")
+  , oscillation_timeout_(0)
+  , oscillation_distance_(0)
+  , recovery_enabled_(true)
+  , behaviors_(behaviors)
+  , action_state_(NONE)
+  , recovery_trigger_(NONE)
+  , dist_to_goal_(std::numeric_limits<double>::infinity())
```
Comment (corot) [inline mbf_abstract_nav/src/move_base_action.cpp:64]: Only this is new; the rest is automatic clang reformatting of the modified code

**Review by spuetz — APPROVED** [review]
lgtm

File: mbf_abstract_nav/src/move_base_action.cpp [hunk 2]
```diff
@@ -121,6 +102,8 @@ void MoveBaseAction::cancel()
 
 void MoveBaseAction::start(GoalHandle &goal_handle)
 {
+  dist_to_goal_ = std::numeric_limits<double>::infinity();
```
Comment (dorezyuk) [inline mbf_abstract_nav/src/move_base_action.cpp:109]: add include limits

File: mbf_abstract_nav/src/move_base_action.cpp [hunk 3]
```diff
@@ -45,22 +45,22 @@
...
+  , oscillation_timeout_(0)
+  , oscillation_distance_(0)
+  , recovery_enabled_(true)
+  , behaviors_(behaviors)
+  , action_state_(NONE)
+  , recovery_trigger_(NONE)
+  , dist_to_goal_(std::numeric_limits<double>::infinity())
+  , replanning_thread_(boost::bind(&MoveBaseAction::replanningThread, this))
 {
 }
```
Comment (dorezyuk) [inline mbf_abstract_nav/src/move_base_action.cpp:68]: we should join the replanning_thread in the destructor

File: mbf_abstract_nav/src/move_base_action.cpp [hunk 4]
```diff
@@ -544,36 +509,57 @@ void MoveBaseAction::actionRecoveryDone(
...
-  }
+    // replan only while following a path and if replanning is enabled (can be disabled by dynamic reconfigure)
+    bool keep_replanning = !replanning_period_.isZero() && action_state_ == EXE_PATH && dist_to_goal_ > 0.1;
 
-  replanning_mtx_.lock();
-  replanning_rate_.sleep();
-  replanning_mtx_.unlock();
+    if (!action_client_get_path_.getState().isDone())
+    {
+      if (action_client_get_path_.waitForResult(update_period))
```
Comment (dorezyuk) [inline mbf_abstract_nav/src/move_base_action.cpp:531]: why do we wait here only so little? did you had issues with shutting down the thread correctly?

File: mbf_abstract_nav/src/move_base_action.cpp [hunk 5]
```diff
@@ -544,36 +509,57 @@ void MoveBaseAction::actionRecoveryDone(
...
+      {
+        actionlib::SimpleClientGoalState state = action_client_get_path_.getState();
+        mbf_msgs::GetPathResultConstPtr result = action_client_get_path_.getResult();
+        if (state == actionlib::SimpleClientGoalState::SUCCEEDED && keep_replanning)
+        {
+          ROS_DEBUG_STREAM_NAMED("move_base", "Replanning succeeded; sending a goal to \"exe_path\" with the new plan");
```
Comment (dorezyuk) [inline mbf_abstract_nav/src/move_base_action.cpp:537]: I think we have here a potential race condition:

When we reach this point keep_replanning might be outdated (it seem to me that this is even likely to happen) and we might then either send a goal to the controller while recovering or when the user/our logic already canceled the current execution

File: mbf_abstract_nav/src/move_base_action.cpp [hunk 6]
```diff
@@ -544,36 +509,57 @@ void MoveBaseAction::actionRecoveryDone(
...
+    {
+      update_period.sleep();
+      continue;
+    }
 
-  ROS_DEBUG_STREAM_NAMED("move_base", "Next replanning cycle, using the \"get_path\" action!");
-  action_client_get_path_.sendGoal(
-      get_path_goal_,
-      boost::bind(&MoveBaseAction::actionGetPathReplanningDone, this, _1, _2)); // replanning
+    if (ros::Time::now() - last_replan_time >= replanning_period_)
```
Comment (dorezyuk) [inline mbf_abstract_nav/src/move_base_action.cpp]: does it make sense to check the state of action_client_get_path_ - I'm thinking that if the user picks a replanning period shorted than the planners response time we will always preempt the planner

File: mbf_abstract_nav/src/move_base_action.cpp [hunk 7]
```diff
@@ -544,36 +509,57 @@ void MoveBaseAction::actionRecoveryDone(
...
+        // keep waiting for planning to complete (we already waited update_period in waitForResult)
+        continue;
+      }
+    }
 
-  if (!replanning_ || action_state_ != EXE_PATH)
-    return; // another chance to stop replannings after waiting for replanning period
+    if (!keep_replanning)
+    {
+      update_period.sleep();
```
Comment (dorezyuk) [inline mbf_abstract_nav/src/move_base_action.cpp:554]: what about using a condition var here with a bigger timeout? we can trigger the condition var every time we set the state to exe_path or change the replanning period

File: mbf_abstract_nav/src/move_base_action.cpp (hunk 6 above)
Comment (dorezyuk) [reply mbf_abstract_nav/src/move_base_action.cpp]: just personal preference but it seems better readable if we make the ifs else ifs:

```
if(!action_client_get_path_.getState().isDone()){...}
else if(!keep_replanning) {...}
else if(ros::Time::now() - last_replan_time >= replanning_period_) {...}
```
in this manner we would also resolve the issue with preempting the planer

File: mbf_abstract_nav/src/move_base_action.cpp (hunk 4 above)
Comment (corot) [reply mbf_abstract_nav/src/move_base_action.cpp:531]: no particular reason,,, 1ms sounds reasonable to react fast to changes on state
will inc x5

File: mbf_abstract_nav/src/move_base_action.cpp (hunk 5 above)
Comment (corot) [reply mbf_abstract_nav/src/move_base_action.cpp:537]: mmm...  yes,,, maybe that's why I set the loop at 1ms.   Probably better now

**Comment by spuetz** [comment]
You are right with the preempting. I like your proposals. This could be done in another PR, or? What do you think @dorezyuk @corot  ?

File: mbf_abstract_nav/src/move_base_action.cpp (hunk 6 above)
Comment (corot) [reply mbf_abstract_nav/src/move_base_action.cpp]: > does it make sense to check the state of action_client_get_path_ - I'm thinking that if the user picks a replanning period shorted than the planners response time we will always preempt the planner

nope,,,  we will continue [here](https://github.com/magazino/move_base_flex/pull/267/files#diff-6b8a12259c3d11b5f68c080c8a6d7a690aecea38015c66366f1d559b16bd4acbR546) and so not send a new goal

Comment (dorezyuk) [reply mbf_abstract_nav/src/move_base_action.cpp]: ah yea you're right - my bad

**Comment by corot** [comment]
> You are right with the preempting. I like your proposals. This could be done in another PR, or? What do you think @dorezyuk @corot ?

I'm done with the requested changes;  just need to test
I'll commit now, before testing, so u can judge

File: mbf_abstract_nav/src/move_base_action.cpp [hunk 8]
```diff
@@ -544,36 +511,53 @@ void MoveBaseAction::actionRecoveryDone(
   }
 }
 
-void MoveBaseAction::actionGetPathReplanningDone(
-    const actionlib::SimpleClientGoalState &state,
-    const mbf_msgs::GetPathResultConstPtr &result)
+bool MoveBaseAction::replanningActive()
```
Comment (dorezyuk) [inline mbf_abstract_nav/src/move_base_action.cpp]: can you make it const?

File: mbf_abstract_nav/src/move_base_action.cpp (hunk 3 above)
Comment (dorezyuk) [reply mbf_abstract_nav/src/move_base_action.cpp:68]: @corot can you fix this - then the pr is fine :)

**Comment by dorezyuk** [comment]
> > You are right with the preempting. I like your proposals. This could be done in another PR, or? What do you think @dorezyuk @corot ?
> 
> I'm done with the requested changes; just need to test
> I'll commit now, before testing, so u can judge

It looks good now - the only thing I could wish for are some tests for this class;)

**Review by dorezyuk — APPROVED** [review]

**Pull request merged**